File: c++/src/hmac_auth.cpp
/*
 ******************************************************************************************************
 Copyright 1998 - 2021 Tencent. All Rights Reserved
 file: hmac_auth.cpp
 date: 2020-04-02
 author: steven
 desc: HMAC authentication implementation
 detail:
 1. The input body is encrypted with SHA-256 and then encoded with base64 to obtain a string of readable encrypted string digest
 2. Add "SHA-256=" before digest, and then spell it into date:<GMT time>\ndigest:<digest>
  Format string 
 3. HMAC encryption of the secret and the content obtained in the second step to obtain the signature sign
 4. Add the following format content to the HTTP request header
  Date: <GMT time>
  Digest: <digest>
  Authorization: hmac username=<user>, algorithm="hmac-sha256", headers="date
 digest", signature=<sign>
 *******************************************************************************************************
*/
#include "hmac_auth.h"

#include <openssl/hmac.h>
#include <openssl/sha.h>

#include <cstring>
#include <ctime>
#include <iostream>
#include <sstream>
#include <string>

HMACAuth::HMACAuth(const std::string &user, const std::string &secret, const std::string &body)
    : user_(user), secret_(secret), body_(body) {}

int HMACAuth::getAuthInfo(AuthInfo* auth_info) {
  std::string date = getGMTTime();
  std::string digest = getDigest(body_);

  std::ostringstream oss;
  oss << "date: " << date << "\ndigest: " << digest;

  std::string sign = hmacEncrypt(secret_, oss.str());

  std::ostringstream auth;
  auth << "hmac username=\"" << user_
       << "\", algorithm=\"hmac-sha256\", headers=\"date digest\", signature=\"" << sign << "\"";

  auth_info->date = date;
  auth_info->digest = digest;
  auth_info->auth = auth.str();

  return 0;
}

std::string HMACAuth::base64Encode(const std::string &text) {
  static unsigned char base64_table[] =
      "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
  char *buf = new char[text.size() * 2];
  char *dst = buf;

  for (size_t i = 0, len = 0; i < text.size(); len += 4) {
    unsigned int c = (unsigned char)text[i++];
    c <<= 8;

    if (i < text.size()) {
      c += (unsigned char)text[i];
    }

    c <<= 8;
    ++i;

    if (i < text.size()) {
      c += (unsigned char)text[i];
    }

    ++i;

    *dst++ = base64_table[(c >> 18) & 0x3f];
    *dst++ = base64_table[(c >> 12) & 0x3f];

    if (i > (text.size() + 1)) {
      *dst++ = '=';
    } else {
      *dst++ = base64_table[(c >> 6) & 0x3f];
    }

    if (i > text.size()) {
      *dst++ = '=';
    } else {
      *dst++ = base64_table[(c >> 0) & 0x3f];
    }
  }

  *dst = '\0';

  return std::string(buf);
}

std::string HMACAuth::sha256Encrypt(const std::string &text) {
  unsigned char hash[SHA256_DIGEST_LENGTH];
  memset(hash, 0, sizeof(hash));

  SHA256_CTX sha256;
  SHA256_Init(&sha256);
  SHA256_Update(&sha256, text.c_str(), text.size());
  SHA256_Final(hash, &sha256);

  // You cannot directly use "ret=(char*)hash;" to assign hash to ret, otherwise there will be an extra character 0x01 at the end
  std::string ret(reinterpret_cast<char *>(hash), sha256.md_len);

  return ret;
}

// 1. Perform sha256 calculation on the body first
// 2. Then base64 encode the string calculated by sha256
// 3. "SHA-256=" should be added to the returned string header
std::string HMACAuth::getDigest(const std::string &body) {
  std::string sha_body = sha256Encrypt(body);
  std::string base64_body = base64Encode(sha_body);

  std::ostringstream oss;
  oss << "SHA-256=" << base64_body;

  return oss.str();
}

std::string HMACAuth::hmacEncrypt(const std::string &secret, const std::string &text) {
  unsigned char out[EVP_MAX_MD_SIZE];
  unsigned int out_size = 0;

  memset(out, 0, sizeof(out));

  HMAC_CTX ctx;
  HMAC_CTX_init(&ctx);

  HMAC_Init_ex(&ctx, secret.c_str(), secret.size(), EVP_sha256(), NULL);
  HMAC_Update(&ctx, reinterpret_cast<const unsigned char *>(text.c_str()), text.size());

  HMAC_Final(&ctx, out, &out_size);
  HMAC_CTX_cleanup(&ctx);

  std::string hmac_str(reinterpret_cast<char *>(out), out_size);
  std::string result =
      base64Encode(hmac_str);  // Do base64 encoding again after HMAC authentication so that the authenticated characters can be printed

  return result;
}

std::string HMACAuth::getGMTTime() {
  struct tm gm_time;
  time_t now = time(0);

  gmtime_r(&now, &gm_time);

  char time_buf[64];
  strftime(time_buf, sizeof(time_buf) - 1, "%a, %d %b %Y %H:%M:%S GMT", &gm_time);

  std::string ret(time_buf);
  return ret;
}

File: python/hmac_auth.py
# -*- coding: utf8 -*-
__author__ = 'Jager'
import sys
import hmac
import hashlib
import base64
import time


class HmacAuth():
     """Generate Hmac (hmac-sha256) authentication header
     :param hmac_user, String, optional, the HMAC account is preset when the class is initialized, which can be overwritten by the function
     :param hmac_secret, String, optional, preset HMAC key when class is initialized, it can be overwritten by function
     :return Class Object
     """
    def __init__(self, hmac_user=None, hmac_secret=None):
        self.hmac_user = hmac_user
        self.hmac_secret = hmac_secret

    def _sha256_digest(self, content):
         """ sha256 calculation content summary
         :param content, String, content
         """
        if sys.version_info.major > 2:
            content_bytes = bytes(content, "utf-8")

        else:
            content_bytes = bytes(content).decode("utf-8")

        content_sha256_digest = hashlib.sha256(content_bytes).digest()
        content_sha256_digest_base64_decode = base64.b64encode(
            content_sha256_digest).decode()
        content_digest = 'SHA-256={}'.format(
            content_sha256_digest_base64_decode)
        return content_digest

    def _hmac_sha256(self, secret, str_to_sign):
         """Generate sha256 encrypted string
         :param secret, String, specify the secret key
         :param str_to_sign, String, assembled data to be signed
         """
        if sys.version_info.major > 2:
            hmac_key = bytes(secret, "utf-8")
            msg_sign = bytes(str_to_sign, "utf-8")

        else:
            hmac_key = bytes(secret)
            msg_sign = bytes(str_to_sign)

        signature = hmac.new(hmac_key, msg_sign,
                             digestmod=hashlib.sha256).digest()
        str_base64 = base64.b64encode(signature).decode()
        return str_base64

    def get_auth_headers(self, hmac_user=None, hmac_secret=None, body=""):
         """Get Hmac authentication header
         :param String, optional, specify the Hmac account, you can override the Hmac account preset by the class
         :param String, optional, specify the hmac key, which can override the Hmac key preset by the class
         :param String, optional, specify the body content of the request, it must be passed when the gateway requires the verification of the body, and a null value is passed in for the Get request
         :param Dict, return Hmac authentication header dictionary
         """
        if not hmac_user:
            hmac_user = self.hmac_user

        if not hmac_secret:
            hmac_secret = self.hmac_secret

        # Generate sha256 encrypted string of body
        body_digest = self._sha256_digest(body)

        # Generate the current GMT time, note that the format cannot be changed, it must be like: Wed, 14 Aug 2019 09:09:28 GMT
        gm_time = time.strftime("%a, %d %b %Y %H:%M:%S GMT", time.gmtime())

        # Assemble the data to be signed
        str_to_sign = "date: {}\ndigest: {}".format(gm_time, body_digest)

        # Generate signature
        signature = self._hmac_sha256(hmac_secret, str_to_sign)

        # Assemble headers
        headers = {}
        headers["Authorization"] = (
            'hmac username="{}", algorithm="hmac-sha256", headers="date digest",'
            'signature="{}"'.format(hmac_user, signature))
        headers["Digest"] = body_digest
        headers["Date"] = gm_time
        return headers
